ipie/addons/thermal/qmc/thermal_afqmc.py

```diff
@@ -0,0 +1,472 @@
+
+                self.tprop_fbias = self.propagator.timer.tfbias
+                self.tprop_update = self.propagator.timer.tupdate
+                self.tprop_vhs = self.propagator.timer.tvhs
+                self.tprop_gemm = self.propagator.timer.tgemm
+                
+                start_clip = time.time()
+                if t > 0:
+                    wbound = self.pcontrol.total_weight * 0.10
+                    xp.clip(self.walkers.weight, a_min=-wbound, a_max=wbound,
+                            out=self.walkers.weight)  # In-place clipping.
+
+                synchronize()
+                self.tprop_clip += time.time() - start_clip
+
+                start_barrier = time.time()
+                if t % self.params.pop_control_freq == 0:
+                    comm.Barrier()
+
+                self.tprop_barrier += time.time() - start_barrier
+                self.tprop += time.time() - start
+                
+                if (t > 0) and (t % self.params.pop_control_freq == 0):
+                    start = time.time()
+                    self.pcontrol.pop_control(self.walkers, comm)
+                    synchronize()
+                    self.tpopc += time.time() - start
+                    self.tpopc_send = self.pcontrol.timer.send_time
+                    self.tpopc_recv = self.pcontrol.timer.recv_time
+                    self.tpopc_comm = self.pcontrol.timer.communication_time
+                    self.tpopc_non_comm = self.pcontrol.timer.non_communication_time
+                
+                # Print cut.
+                #self.estimators.compute_estimators(
+                #        self.hamiltonian, self.trial, self.walkers)
+                #self.estimators.print_cut(
+                #        comm, step // self.params.num_steps_per_block, t, self.accumulators)
+
+            # Accumulate weight, hybrid energy etc. across block.
+            start = time.time()
+            self.accumulators.update(self.walkers)
+            self.testim += time.time() - start
+
+            # Calculate estimators.
+            start = time.time()
+            if step % self.params.num_steps_per_block == 0:
+                self.estimators.compute_estimators(
+                        self.hamiltonian, self.trial, self.walkers)
+
+                self.estimators.print_block(
+                    comm, step // self.params.num_steps_per_block, self.accumulators)
+                self.accumulators.zero()
+
+            synchronize()
+            self.testim += time.time() - start
+
+            if step < neqlb_steps:
+                eshift = self.accumulators.eshift
+
+            else:
+                eshift += self.accumulators.eshift - eshift
+
+            self.walkers.reset(self.trial) # Reset stack, weights, phase.
+
+            synchronize()
+            self.tpath += time.time() - start_path
+
+
+    def finalise(self, verbose=False):
```

Comment:
If you didn't modify the timers just delete this and allow inheritance to do its job.